src/layers/regularizers/instance_norm.cpp

```diff
@@ -0,0 +1,237 @@
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#define LBANN_INSTANCE_NORM_LAYER_INSTANTIATE
+#include "lbann/layers/regularizers/instance_norm.hpp"
+
+namespace lbann {
+
+namespace {
+
+/** @brief Forward prop */
+template <typename TensorDataType>
+void fp_impl(lbann_comm& comm,
+             El::Int num_channels,
+             El::Int channel_size,
+             TensorDataType epsilon,
+             const El::AbstractDistMatrix<TensorDataType>& input,
+             El::AbstractDistMatrix<TensorDataType>& output,
+             El::AbstractDistMatrix<TensorDataType>& statistics) {
+
+  // Local matrices
+  using LocalMat = El::Matrix<TensorDataType, El::Device::CPU>;
+  const auto& local_input = dynamic_cast<const LocalMat&>(input.LockedMatrix());
+  auto& local_output = dynamic_cast<LocalMat&>(output.Matrix());
+  auto& local_statistics = dynamic_cast<LocalMat&>(statistics.Matrix());
+  auto local_means = El::View(local_statistics,
+                              El::IR(0, num_channels), El::ALL);
+  auto local_vars = El::View(local_statistics,
+                             El::IR(num_channels, 2*num_channels), El::ALL);
+
+  // Dimensions
+  const El::Int local_mini_batch_size = local_input.Width();
+
+  // Trivial case if channel size is 1
+  // Note: Output is constant.
+  if (channel_size <= 1) {
+    El::Zero(output);
+    return;
+  }
+
+  // Compute sums
+  El::Zero(statistics);
+  LBANN_OMP_PARALLEL_FOR_COLLAPSE2
+  for (El::Int k = 0; k < local_mini_batch_size; ++k) {
+    for (El::Int j = 0; j < num_channels; ++j) {
+      auto& sum = local_means(j,k);
+      auto& sqsum = local_vars(j,k);
+      for (El::Int i = 0; i < channel_size; ++i) {
+        const auto& x = local_input(i+j*channel_size,k);
+        sum += x;
+        sqsum += x * x;
+      }
+    }
+  }
+
+  // Compute statistics from sums
+  //   mean = sum(x_i) / n
+  //   var = ( sum(x_i^2)/n - mean^2 ) * n/(n-1)
+  LBANN_OMP_PARALLEL_FOR_COLLAPSE2
+  for (El::Int k = 0; k < local_mini_batch_size; ++k) {
```

Comment:
I think this loop and the next one can be fused together?

Reply:
Good point. We can do something similar in the GPU implementation without too much effort, but it will be ugly. I'm thinking that reducing runtime by 1/3 might be worth it though...

Reply:
Yes, I think so. Can do it here or in a separate PR. 🚀

Reply:
Done.